
/*___________________________________________________
 |  _____                       _____ _ _       _    |
 | |  __ \                     |  __ (_) |     | |   |
 | | |__) |__ _ __   __ _ _   _| |__) || | ___ | |_  |
 | |  ___/ _ \ '_ \ / _` | | | |  ___/ | |/ _ \| __| |
 | | |  |  __/ | | | (_| | |_| | |   | | | (_) | |_  |
 | |_|   \___|_| |_|\__, |\__,_|_|   |_|_|\___/ \__| |
 |                   __/ |                           |
 |  GNU/Linux based |___/  Multi-Rotor UAV Autopilot |
 |___________________________________________________|
 
 Message Format for interfacing OPCD

 Copyright (C) 2014 Tobias Simon, Integrated Communication Systems Group, TU Ilmenau

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details. */


message Value
{
   optional string str_val = 1;
   optional int32 int_val = 2;
   optional double dbl_val = 3;
   optional bool bool_val = 4;
}


message CtrlReq
{
   enum Type
   {
      GET = 0; /* get parameter */
      SET = 1; /* set parameter to value */
      PERSIST = 2; /* persist all parameters */
   }

   required Type type = 1;
   optional string id = 2;
   optional Value val = 3;
}


message Pair
{
   required string id = 1;
   required Value val = 2;
}


message CtrlRep
{
   enum Status
   {
      OK = 0;
      PARAM_UNKNOWN = 1;
      TYPE_MISMATCH = 2;
      MALFORMED_ID = 3;
      IO_ERROR = 4;
   }

   required Status status = 1;
   repeated Pair pairs = 2;
}


/* 
 * NOTE: there is no dedicated protobuf message for param publishing using ZMQ pub/sub;
 *       this data consists of a PLAIN null-terminated string, followed by a "Value" protobuf message
 *
 * pseudo-message PubData
 * {
 *    required string id = 1;
 *    required Value val = 2;
 * }
 */

